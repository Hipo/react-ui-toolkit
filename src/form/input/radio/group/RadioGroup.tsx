import "./_radio-group.scss";

import React from "react";
import classNames from "classnames";

import RadioInput, {RadioInputItem, RadioInputProps} from "../Radio";

export interface RadioGroupProps {
  items: RadioInputItem[];
  selectedItem: null | RadioInputItem;
  onSelect: RadioInputProps["onSelect"];
  customClassName?: string;
  isDisabled?: boolean;
  testid?: string;
}

function RadioGroup({
  items,
  testid,
  onSelect,
  selectedItem,
  customClassName,
  isDisabled
}: RadioGroupProps) {
  const radioGroupClassName = classNames("radio-group", customClassName);

  return (
    <ul data-testid={testid} className={radioGroupClassName}>
      {items.map((item) => (
        <li
          key={item.id}
          className={classNames("radio-group-item", item.customClassName)}
          data-testid={`${testid}.item`}>
          <RadioInput
            isDisabled={isDisabled || item.isDisabled}
            item={item}
            onSelect={onSelect}
            isSelected={Boolean(selectedItem && selectedItem.id === item.id)}
          />
        </li>
      ))}
    </ul>
  );
}

export default RadioGroup;
