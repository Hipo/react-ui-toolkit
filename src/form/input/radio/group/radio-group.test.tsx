import React from "react";
import {render, screen} from "@testing-library/react";
import "@testing-library/jest-dom";

import RadioGroup, {RadioGroupProps} from "./RadioGroup";
import {testA11y} from "../../../../core/utils/test/testUtils";

describe("<RadioGroup />", () => {
  const radioGroupItems: RadioGroupProps["items"] = [
    {
      id: "radio-input-1",
      content: <p data-testid={"radio-input.content-0"}>{"Test 1"}</p>,
      inputProps: {htmlFor: "radio-input", name: "radio-input", value: "test.1"}
    },
    {
      id: "radio-input-2",
      content: <p data-testid={"radio-input.content-1"}>{"Test 2"}</p>,
      inputProps: {htmlFor: "radio-input", name: "radio-input", value: "test.2"}
    },
    {
      id: "radio-input-3",
      content: <p data-testid={"radio-input.content-2"}>{"Test 3"}</p>,
      inputProps: {htmlFor: "radio-input", name: "radio-input", value: "test.3"}
    }
  ];

  const defaultRadioGroupProps: RadioGroupProps = {
    testid: "radio-group",
    items: radioGroupItems,
    onSelect: jest.fn(),
    selectedItem: null
  };

  it("should render correctly", () => {
    render(<RadioGroup {...defaultRadioGroupProps} />);
  });

  it("should pass a11y test", async () => {
    const {container} = render(<RadioGroup {...defaultRadioGroupProps} />);

    await testA11y(container, {rules: {"duplicate-id-aria": {enabled: false}}});
  });

  it("should render items correctly", () => {
    render(<RadioGroup {...defaultRadioGroupProps} />);

    expect(screen.getAllByRole("listitem").length).toEqual(radioGroupItems.length);

    for (let index = 0; index < radioGroupItems.length; index++) {
      const radioInputContent = screen.getByTestId(`radio-input.content-${index}`);

      expect(screen.getByTestId(`radio-group.item-${index}`)).toContainElement(
        radioInputContent
      );
    }
  });

  it("should add checked attribute to selectedItem", () => {
    const {rerender} = render(<RadioGroup {...defaultRadioGroupProps} />);

    for (let index = 0; index < radioGroupItems.length; index++) {
      expect(
        screen.getByDisplayValue(radioGroupItems[index].inputProps.value)
      ).not.toBeChecked();
    }

    rerender(
      <RadioGroup {...defaultRadioGroupProps} selectedItem={radioGroupItems[1]} />
    );

    expect(screen.getByDisplayValue(radioGroupItems[1].inputProps.value)).toBeChecked();
  });
});
