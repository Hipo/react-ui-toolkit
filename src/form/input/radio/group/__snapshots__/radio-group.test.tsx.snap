// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup /> should match snapshot 1`] = `
<ul
  className="list radio-group"
  data-testid="radio-group"
>
  <li
    className="list-item radio-group__item"
  >
    <label
      className="radio-input-label"
      data-testid="radio-group.item-0"
      htmlFor="radio-input"
    >
      <input
        checked={false}
        className="radio-input"
        id="radio-input"
        name="radio-input"
        onChange={[Function]}
        type="radio"
        value="test.1"
      />
      <span
        className="radio-input-label__icon"
      />
      <p
        data-testid="radio-input.content-0"
      >
        Test 1
      </p>
    </label>
  </li>
  <li
    className="list-item radio-group__item"
  >
    <label
      className="radio-input-label"
      data-testid="radio-group.item-1"
      htmlFor="radio-input"
    >
      <input
        checked={false}
        className="radio-input"
        id="radio-input"
        name="radio-input"
        onChange={[Function]}
        type="radio"
        value="test.2"
      />
      <span
        className="radio-input-label__icon"
      />
      <p
        data-testid="radio-input.content-1"
      >
        Test 2
      </p>
    </label>
  </li>
  <li
    className="list-item radio-group__item"
  >
    <label
      className="radio-input-label"
      data-testid="radio-group.item-2"
      htmlFor="radio-input"
    >
      <input
        checked={false}
        className="radio-input"
        id="radio-input"
        name="radio-input"
        onChange={[Function]}
        type="radio"
        value="test.3"
      />
      <span
        className="radio-input-label__icon"
      />
      <p
        data-testid="radio-input.content-2"
      >
        Test 3
      </p>
    </label>
  </li>
</ul>
`;
